### ISSUE TITLE
Protocol __getattribute__ not checked against class members

### ISSUE THREAD

**Issue by mrRachar** [astral-sh/ty#3786]
### Summary

It is possible to bypass attribute type checking pretty easily using a protocol:

```py
from typing import Protocol

class IntName(Protocol):
    def __getattribute__(self, name: str, /) -> int: ...

class Person:
    name: str

def get_name_as_int(person: IntName) -> int:
    return person.name

get_name_as_int(Person())  # would expect invalid argument type diagnostic
```

_[playground](https://play.ty.dev/c40894dd-9032-4b82-b6fc-03175930c5b1)_

This is because there is no checking that a protocol's `__getattribute__` type is valid for the fields of the prospective subtype class, instead `object`'s definition is checked which is easily satisfiable. I'm not sure whether this is considered a safety problem today: it depends on whether you consider these field accesses to be equivalent to a `__getattribute__` call (where this is fine since `object`'s return type is permissive). It would be nice if this issue were caught; keeping field accesses as safe seems important regardless of if the default `__getattribute__` method returns `Any`.

In general, it looks like it would only be safe to type something with a Protocol with  `__getattribute__` if (a) the prospective subtype overrides `object.__getattribute__`, or (b)  the Protocol's `__getattribute__` name parameter and return type combinations are valid for the prospective subtype. This becomes something more useful with literal types (where this literal might come from some typevar as part of some wider library code):

```py
class StringAttribute[N: str](Protocol):
    def __getattribute__(self, name: N, /) -> str: ...

def get_name(person: StringAttribute[Literal["name"]]) -> str:
    return person.name

get_name(Person())
```

### Version

041a4e1d9

**Comment by mrRachar** [comment]
I have [a branch](https://github.com/astral-sh/ruff/commit/ddc0c290e5332e369529fee9fbf75a0a1022c613#diff-4e6037bcb844d660fc41a886c1cdeacb32d53ccc4b88438b993e44e18f1a7582R817) up with a change and tests that takes a similar approach to `TypedDict` in that it synthesises a `__getattribute__`, and only doing so during Protocol member checking to avoid changing the behaviour of `__getattribute__` otherwise to keep the impact minimal for now, but I'm not sure the approach that be best here. Happy to put up a PR with it if it seems reasonable; I'm putting up an issue first since this does feel like more of a feature request than a bugfix given its scope.

One advantage of fixing this also allows for a pattern matching approach against types, which allows you to deconstruct them by field and type a `getattr` equivalent:

```py
from typing import Protocol, Literal

class HasAttr[A: str, T](Protocol):
    def __getattribute__(self, attribute: A, /) -> T: ...  # ty: ignore[invalid-method-override]

class IntStrPair:
    left: int
    right: str

def my_getattr[A: str, T](object: HasAttr[A, T], attr: A) -> T:
    return getattr(object, attr)

reveal_type(my_getattr(IntStrPair(), "left"))  # revealed: int
reveal_type(my_getattr(IntStrPair(), "right"))  # revealed: str
```

You can see this passing [in an mdtest added in the branch](https://github.com/astral-sh/ruff/commit/ddc0c290e5332e369529fee9fbf75a0a1022c613#diff-19ac4a4d9337821d0bbfb544ab4e3fda87df59c6ea06065b77d3fcf5cf682ec2R2254-R2262)

**Comment by sharkdp** [comment]
Thank you for your report.

No type checker supports this right now, and we're unlikely to prioritize this, especially if it's unclear if this wouldn't introduce any problems. But feel free to open a (draft) PR with your changes, then we could see the ecosystem impact of such a change.
